src/content/docs/tutorials/counter-app/basic-app.md

```diff
@@ -0,0 +1,352 @@
+`saturating_add` to avoid panics like this.
+
+:::
+
+Run the tests.
+
+```shell title="run tests"
+cargo test
+```
+
+You should see:
+
+```text title="test output"
+running 4 tests
+test tests::handle_key_event ... ok
+test tests::handle_key_event_invalid ... ok
+test tests::handle_key_event_integer - should panic ... ok
+test tests::render_frame ... ok
+
+test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
+```
+
+## The Finished App
+
+Putting this altogether, you should now have the following two files.
+
+<details>
+<summary>src/main.rs</summary>
+
+```rust
+{{#include @code/counter-app/src/main.rs }}
+```
+
+</details>
+
+<details>
+<summary>src/tui.rs</summary>
+
+```rust
+{{#include @code/counter-app/src/tui.rs }}
+```
+
+</details>
+
+## Running the app
+
+Make sure you save all the files and then run the app using:
+
+```shell title="run the app"
+cargo run
+```
+
+You will see the following UI:
+
+![basic-app demo](./basic-app.gif)
+
+Press 'j', 'k', Up and Down to interact with the counter. Press 'q' or Escape to quit.
+
+## Handling Panics
+
+Experiment to see what happens when the application panics. The application has an intentional bug
+where it uses `u8` for the counter field, but doesn't guard against decrementing this below 0. Run
+the app and press 'j' or the Down key.
+
+![basic-app-panic demo](./basic-app-panic.png)
+
+To get more information about where the error occurred, add `RUST_BACKTRACE=full` before the
+command.
+
+![basic-app-panic-full demo](./basic-app-panic-full.png)
+
+## Handling Errors
+
+Experiment to see what happens when the application returns an unhandled error as a result. The
+app will cause this to happen when any key other than one expected is pressed. Run the app and
+press any key except the ones configured.
+
+![basic-app-error demo](./basic-app-error.png)
+
+To get more information about where the error occurred, add `RUST_BACKTRACE=full` before the
+command.
+
+![basic-app-error-full demo](./basic-app-error-full.png)
+
+## Conclusion
+
+By understanding the structure and components used in this simple counter application, you are set
+up to explore crafting more intricate terminal-based interfaces using `ratatui`.
+
+[backends]: /concepts/backends/
+[alternate screen]: /concepts/backends/alternate-screen/
+[raw mode]: /concepts/backends/raw-mode/
+[hello world tutorial]: /tutorials/hello-world/
+[`color_eyre`]: <https://crates.io/crates/color_eyre>
+[Crossterm events module]: https://docs.rs/crossterm/latest/crossterm/event/index.html
+[`Paragraph`]: https://docs.rs/ratatui/latest/ratatui/widgets/struct.Paragraph.html
+[`Block`]: https://docs.rs/ratatui/latest/ratatui/widgets/block/struct.Block.html
+[`TestBackend`]: https://docs.rs/ratatui/latest/ratatui/backend/struct.TestBackend.html
+[`event::read`]: https://docs.rs/crossterm/latest/crossterm/event/fn.read.html
+[`event::poll`]: https://docs.rs/crossterm/latest/crossterm/event/fn.poll.html
```

Comment:
I like the links at the bottom of the file like this. Can we add a note to the developer guide that we should write documentation like this?

Reply:
Links like this would work at the bottom of each section as well right? I have a minor preference for writing them at the bottom of each section so that markdown is easy to review / edit but not a super big deal.

Reply:
Yep to both questions. The best part about it is that it doesn't interrupt the article flow when you're reading. 
Mostly I put them at the bottom because sometimes there can be links used multiple times and it's easier to see that / find and correct links if they're all in one place. I'm not fazed either way though, so happy to change it / leave it and make that the standard approach in the future.

Reply:
That makes sense for links that are used multiple times! I like standardizing it for all links. I wish prettier just did that transformation when formatting markdown files.